### ISSUE TITLE
go_genrule's $GOROOT missing srcs and bin directory with race detector enabled

### ISSUE THREAD

**Issue by ixdy** [kubernetes/repo-infra#86]
For some reason, updating to go1.11+ breaks `openapi-gen`, but only when `--features=race` is enabled:

```console
$ bazel build --features=race pkg/generated/openapi/...

INFO: Build options have changed, discarding analysis cache.
INFO: Analysed 5 targets (0 packages loaded).
INFO: Found 5 targets...
ERROR: /usr/local/google/home/jgrafton/code/kubernetes/src/k8s.io/kubernetes/pkg/generated/openapi/BUILD:9:1: error executing shell command: '/bin/bash -c set -e
export GO_GENRULE_EXECROOT=$(pwd)
export GOPATH=$GO_GENRULE_EXECROOT/bazel-out/k8-fastbuild/bin/pkg/generated/openapi/zz_generated.openapi~gopath
export GOROOT=$GO_GENRULE_EXECR...' failed (Exit 1): bash failed: error executing command
  (cd /usr/local/google/home/jgrafton/.cache/bazel/_bazel_jgrafton/75e5737b0f848bf1b7c3ab3303c8bf72/execroot/__main__ && \
  exec env - \
    CGO_ENABLED=1 \
    GOARCH=amd64 \
    GOOS=linux \
    GOROOT=bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_race_stripped/stdlib% \
    GOROOT_FINAL=GOROOT \
    PATH=bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_race_stripped/stdlib%/bin:/bin:/usr/bin \
  /bin/bash -c 'set -e
export GO_GENRULE_EXECROOT=$(pwd)
export GOPATH=$GO_GENRULE_EXECROOT/bazel-out/k8-fastbuild/bin/pkg/generated/openapi/zz_generated.openapi~gopath
export GOROOT=$GO_GENRULE_EXECROOT/bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_race_stripped/stdlib%
cd $GOPATH/src/k8s.io/kubernetes/; $GO_GENRULE_EXECROOT/bazel-out/host/bin/vendor/k8s.io/kube-openapi/cmd/openapi-gen/linux_amd64_race_stripped/openapi-gen --v 1 --logtostderr --go-header-file $GO_GENRULE_EXECROOT/hack/boilerplate/boilerplate.go.txt --output-file-base zz_generated.openapi --output-package k8s.io/kubernetes/pkg/generated/openapi --report-filename tmp_api_violations.report --input-dirs k8s.io/kubernetes/cmd/cloud-controller-manager/app/apis/config/v1alpha1,k8s.io/kubernetes/pkg/apis/abac/v0,k8s.io/kubernetes/pkg/apis/abac/v1beta1,k8s.io/kubernetes/pkg/version,k8s.io/kubernetes/vendor/k8s.io/api/admission/v1beta1,k8s.io/kubernetes/vendor/k8s.io/api/admissionregistration/v1alpha1,k8s.io/kubernetes/vendor/k8s.io/api/admissionregistration/v1beta1,k8s.io/kubernetes/vendor/k8s.io/api/apps/v1,k8s.io/kubernetes/vendor/k8s.io/api/apps/v1beta1,k8s.io/kubernetes/vendor/k8s.io/api/apps/v1beta2,k8s.io/kubernetes/vendor/k8s.io/api/authentication/v1,k8s.io/kubernetes/vendor/k8s.io/api/authentication/v1beta1,k8s.io/kubernetes/vendor/k8s.io/api/authorization/v1,k8s.io/kubernetes/vendor/k8s.io/api/authorization/v1beta1,k8s.io/kubernetes/vendor/k8s.io/api/autoscaling/v1,k8s.io/kubernetes/vendor/k8s.io/api/autoscaling/v2beta1,k8s.io/kubernetes/vendor/k8s.io/api/autoscaling/v2beta2,k8s.io/kubernetes/vendor/k8s.io/api/batch/v1,k8s.io/kubernetes/vendor/k8s.io/api/batch/v1beta1,k8s.io/kubernetes/vendor/k8s.io/api/batch/v2alpha1,k8s.io/kubernetes/vendor/k8s.io/api/certificates/v1beta1,k8s.io/kubernetes/vendor/k8s.io/api/coordination/v1beta1,k8s.io/kubernetes/vendor/k8s.io/api/core/v1,k8s.io/kubernetes/vendor/k8s.io/api/events/v1beta1,k8s.io/kubernetes/vendor/k8s.io/api/extensions/v1beta1,k8s.io/kubernetes/vendor/k8s.io/api/imagepolicy/v1alpha1,k8s.io/kubernetes/vendor/k8s.io/api/networking/v1,k8s.io/kubernetes/vendor/k8s.io/api/policy/v1beta1,k8s.io/kubernetes/vendor/k8s.io/api/rbac/v1,k8s.io/kubernetes/vendor/k8s.io/api/rbac/v1alpha1,k8s.io/kubernetes/vendor/k8s.io/api/rbac/v1beta1,k8s.io/kubernetes/vendor/k8s.io/api/scheduling/v1alpha1,k8s.io/kubernetes/vendor/k8s.io/api/scheduling/v1beta1,k8s.io/kubernetes/vendor/k8s.io/api/settings/v1alpha1,k8s.io/kubernetes/vendor/k8s.io/api/storage/v1,k8s.io/kubernetes/vendor/k8s.io/api/storage/v1alpha1,k8s.io/kubernetes/vendor/k8s.io/api/storage/v1beta1,k8s.io/kubernetes/vendor/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1,k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/api/resource,k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/apis/meta/v1,k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/apis/meta/v1beta1,k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/apis/testapigroup/v1,k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime,k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/intstr,k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/version,k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/apis/audit/v1,k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/apis/audit/v1alpha1,k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/apis/audit/v1beta1,k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/apis/example/v1,k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/apis/example2/v1,k8s.io/kubernetes/vendor/k8s.io/client-go/pkg/apis/clientauthentication/v1alpha1,k8s.io/kubernetes/vendor/k8s.io/client-go/pkg/apis/clientauthentication/v1beta1,k8s.io/kubernetes/vendor/k8s.io/client-go/pkg/version,k8s.io/kubernetes/vendor/k8s.io/csi-api/pkg/apis/csi/v1alpha1,k8s.io/kubernetes/vendor/k8s.io/kube-aggregator/pkg/apis/apiregistration/v1,k8s.io/kubernetes/vendor/k8s.io/kube-aggregator/pkg/apis/apiregistration/v1beta1,k8s.io/kubernetes/vendor/k8s.io/kube-controller-manager/config/v1alpha1,k8s.io/kubernetes/vendor/k8s.io/kube-proxy/config/v1alpha1,k8s.io/kubernetes/vendor/k8s.io/kube-scheduler/config/v1alpha1,k8s.io/kubernetes/vendor/k8s.io/kubelet/config/v1beta1,k8s.io/kubernetes/vendor/k8s.io/metrics/pkg/apis/custom_metrics/v1beta1,k8s.io/kubernetes/vendor/k8s.io/metrics/pkg/apis/custom_metrics/v1beta2,k8s.io/kubernetes/vendor/k8s.io/metrics/pkg/apis/external_metrics/v1beta1,k8s.io/kubernetes/vendor/k8s.io/metrics/pkg/apis/metrics/v1alpha1,k8s.io/kubernetes/vendor/k8s.io/metrics/pkg/apis/metrics/v1beta1 && cp $GOPATH/src/k8s.io/kubernetes/pkg/generated/openapi/zz_generated.openapi.go $GO_GENRULE_EXECROOT/bazel-out/k8-fastbuild/genfiles/pkg/generated/openapi/zz_generated.openapi.go && rm tmp_api_violations.report')

Use --sandbox_debug to see verbose messages from the sandbox
E1003 17:47:36.869706       4 openapi.go:365] Error when generating: Duration, Duration invalid type
2018/10/03 17:47:36 OpenAPI code generation error: Failed executing generator: some packages had errors:
cannot generate spec for type invalid type
INFO: Elapsed time: 13.249s, Critical Path: 12.61s
INFO: 1 process: 1 linux-sandbox.
FAILED: Build did NOT complete successfully
```

Digging in, this appears to be because we are setting `GOROOT` to `go_context().root` instead of `go_context().sdk.root_file.dirname`.

When there aren't any features enabled, `.root` and `.sdk.root_file.dirname` point to the same location, the precompiled go stdlib and sources.

When we enable a feature like the race detector, `rules_go` compiles the stdlib, and then `go_context().root` points to this compiled library. There's no tools or sources in this directory, though, just a `pkg` directory, which causes problems for `openapi-gen`. (I have no idea how this worked before, since I see the same behavior when using `go1.10.4`.)

Ideally, I'd like to have `GOROOT` point at a directory that contains the downloaded, precompiled `go_sdk` as well as the feature-compiled sdk, but I'm not sure if that's possible with `rules_go` currently. Or am I misunderstanding this completely?

cc @BenTheElder @cblecker @jayconrod

**Comment by jayconrod** [comment]
> Digging in, this appears to be because we are setting GOROOT to go_context().root instead of go_context().sdk.root_file.dirname.

> When there aren't any features enabled, .root and .sdk.root_file.dirname point to the same location, the precompiled go stdlib and sources.

> When we enable a feature like the race detector, rules_go compiles the stdlib, and then go_context().root points to this compiled library. There's no tools or sources in this directory, though, just a pkg directory, which causes problems for openapi-gen. (I have no idea how this worked before, since I see the same behavior when using go1.10.4.)

This summary is correct. `go.root` and `go.env["GOROOT"]` point to a directory containing compiled packages for the target platform. `go.sdk.root_file.dirname` points to the directory containing the downloaded SDK, which includes compiled packages for the host platform. These are the same by default, but when a non-default configuration is used, the standard library is compiled somewhere else. In earlier versions of rules_go (at or before 0.13.0 I think), the standard library was always recompiled somewhere else, and it included the `src` directory, but copying and compiling these files took a long time, especially since the standard library is an input for most actions.

If you need a full `GOROOT` with a `src` directory and a `pkg` directory compiled in the target configuration, probably the best way to do that is to construct it with symlinks to subdirectories in the above locations.

I'm not sure what changed in Go 1.11 to expose this, but it seems like a bug in `go_genrule` in any case.

**Comment by ixdy** [comment]
fixed in #87.
